File: src/main/java/es/pharmashop/persistence/sqlite/entity/Product.java
package es.pharmashop.persistence.sqlite.entity;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.util.StringUtils;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.Table;

@Data
@Entity
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "product")
public class Product {
  @Id
  @Column(name = "ean")
  String ean;

  @Column(name = "unique_code")
  String uniqueCode;

  @Column(name = "cn")
  String cn;

  @Column(name = "name")
  String name;

  @Column(name = "revision")
  Integer revision;

  public boolean validAndReviewed() {
    return StringUtils.hasText(uniqueCode) && revision == 1;
  }

  public Product clone() {
    return Product.builder()
      .ean(ean)
      .uniqueCode(uniqueCode)
      .cn(cn)
      .name(name)
      .revision(revision)
      .build();
  }
}

File: front/src/app/domain/Product.ts
export interface IProduct {
  uniqueCode: string;
  cn: string;
  ean: string;
  name: string;
  revision: boolean;
}

export class Product {
  public uniqueCode: string;
  public cn: string;
  public ean: string;
  public name: string;
  public revision: boolean;

  static empty(): Product {
    return new Product({
      uniqueCode: null,
      cn: null,
      ean: null,
      name: null,
      revision: false
    });
  }

  constructor({uniqueCode, cn, ean, name, revision}: IProduct) {
    this.uniqueCode = uniqueCode;
    this.cn = cn;
    this.ean = ean;
    this.name = name;
    this.revision = revision;
  }

  isValid(): boolean {
    if (!this.uniqueCode) {
      return false;
    }

    if (!this.cn) {
      return false;
    }

    return `${this.ean}`.length === 13;
  }

  isNotValid(): boolean {
    return !this.isValid();
  }

  clone() {
    return new Product(this);
  }
}
